### PR TITLE
.delegatecall() should always return a boolean of execution status

### LINKED ISSUE #2678: .delegatecall() should always return the execution outcome
Reported in https://github.com/paritytech/parity/issues/6201.

### PR DESCRIPTION
**Pull request by axic**
Fixes #2678.

### CODE REVIEW

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 1]
```diff
@@ -9816,6 +9816,42 @@ BOOST_AUTO_TEST_CASE(inlineasm_empty_let)
...
+			function set(uint _value) external {
+				value = _value;
+			}
+			function get() external constant returns (uint) {
+				return value;
+			}
+			function get_delegated() external constant returns (bool) {
+				return this.delegatecall(bytes4(sha3("get()")));
+			}
+			function get_delegated(address _address) external constant returns (bool) {
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Can you also add a function that causes an exception?

Comment (axic) [reply test/libsolidity/SolidityEndToEndTest.cpp]: Added a case with assert.

**Comment by chriseth** [comment]
Please also add an entry in the bug list json.

File: docs/bugs.json [hunk 2]
```diff
@@ -1,14 +1,22 @@
 [
     {
+        "name": "DelegateCallReturnValue",
+        "summary": "The low-level .delegatecall() took the returned value as a boolean instead of the execution outcome.",
```
Comment (chriseth) [inline docs/bugs.json]: I think the other summaries are written in present tense.
Also note that the returned value can be garbage, if the called contract does not return anything.

File: docs/bugs.json [hunk 3]
```diff
@@ -1,14 +1,22 @@
...
+        "description": "If the destination would return a value, which interpreted as a boolean means false, the return value of .delegatecall() would be false, despite successful execution.",
```
Comment (chriseth) [inline docs/bugs.json]: Why would it mean false?

Comment (axic) [reply docs/bugs.json]: It meant: *If it means false when interpreted as a boolean*.

Comment (chriseth) [reply docs/bugs.json]: Ah, then I think we should write that without a comma. But what about writing "If the called contract returns zero, ..."?

Comment (chriseth) [reply docs/bugs.json]: (or something starting with 32 zero bytes)

File: docs/bugs.json (hunk 2 above)
Comment (axic) [reply docs/bugs.json]: Changed to present.

File: docs/bugs.json [hunk 4]
```diff
@@ -1,14 +1,22 @@
 [
     {
+        "name": "DelegateCallReturnValue",
+        "summary": "The low-level .delegatecall() takes the returned value as a boolean instead of the execution outcome.",
```
Comment (chriseth) [inline docs/bugs.json]: What about `... does not return the execution outcome but converts the value returned by the function call to a boolean instead.`

File: docs/bugs.json [hunk 5]
```diff
@@ -1,14 +1,22 @@
...
+        "description": "If the destination returns a value or garbage, which if interpreted as a boolean means false, the return value of .delegatecall() is false, despite successful execution.",
```
Comment (chriseth) [inline docs/bugs.json]: This happens even if the destination does not return anything (not even garbage).

What about: `The return value of the low-level .delegatecall() function is taken from a position in memory, where the call data or the return data resides. This value is interpreted as a boolean and put onto the stack. This means if the called function returns at least 32 zero bytes, .delegatecall() returns false even if the call was successuful.`

File: docs/bugs.json (hunk 4 above)
Comment (axic) [reply docs/bugs.json]: Sounds good.

**Comment by axic** [comment]
@chriseth should be ready

**Review by chriseth — APPROVED** [review]

**Pull request merged**